domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -161,101 +192,305 @@ class TopicListController @Inject constructor(
+      topicProgressList.sortedByDescending { it.lastPlayedTimestamp }
+
+    sortedTopicProgressList.forEach { topicProgress ->
       val topic = topicController.retrieveTopic(topicProgress.topicId)
+
       topicProgress.storyProgressMap.values.forEach { storyProgress ->
         val storyId = storyProgress.storyId
         val story = topicController.retrieveStory(topic.topicId, storyId)
 
-        val completedChapterProgressList =
-          storyProgress.chapterProgressMap.values
-            .filter { chapterProgress ->
-              chapterProgress.chapterPlayState ==
-                ChapterPlayState.COMPLETED
-            }
-            .sortedByDescending { chapterProgress -> chapterProgress.lastPlayedTimestamp }
-
-        val lastCompletedChapterProgress: ChapterProgress? =
+        val completedChapterProgressList = getCompletedChapterProgressList(storyProgress)
+        val mostRecentCompletedChapterProgress: ChapterProgress? =
           completedChapterProgressList.firstOrNull()
 
-        val startedChapterProgressList =
-          storyProgress.chapterProgressMap.values
-            .filter { chapterProgress ->
-              chapterProgress.chapterPlayState ==
-                ChapterPlayState.STARTED_NOT_COMPLETED
-            }
-            .sortedByDescending { chapterProgress -> chapterProgress.lastPlayedTimestamp }
-
+        val startedChapterProgressList = getStartedChapterProgressList(storyProgress)
         val recentlyPlayerChapterProgress: ChapterProgress? =
           startedChapterProgressList.firstOrNull()
-        if (recentlyPlayerChapterProgress != null) {
-          val recentlyPlayerChapterSummary: ChapterSummary? =
-            story.chapterList.find { chapterSummary ->
-              recentlyPlayerChapterProgress.explorationId == chapterSummary.explorationId
-            }
-          if (recentlyPlayerChapterSummary != null) {
-            val numberOfDaysPassed =
-              (Date().time - recentlyPlayerChapterProgress.lastPlayedTimestamp) / ONE_DAY_IN_MS
-            val promotedStory = createPromotedStory(
+
+        checkIfStoryIsCompleted(topic.topicId, mostRecentCompletedChapterProgress, story)
+
+        when {
+          recentlyPlayerChapterProgress != null -> {
+            createOngoingStoryListBasedOnRecentlyPlayed(
               storyId,
+              story,
+              recentlyPlayerChapterProgress,
+              completedChapterProgressList,
               topic,
-              completedChapterProgressList.size,
-              story.chapterCount,
-              recentlyPlayerChapterSummary.name,
-              recentlyPlayerChapterSummary.explorationId
-            )
-            if (numberOfDaysPassed < ONE_WEEK_IN_DAYS) {
-              ongoingStoryListBuilder.addRecentStory(promotedStory)
-            } else {
-              ongoingStoryListBuilder.addOlderStory(promotedStory)
+              completedStoryTopicId
+            ).let {
+              Pair(it.first, it.second)
```

Comment:
I'm having a lot of difficulty following the Pair logic. Given my earlier comment regarding doing the merging logic in the UI layer, could we just have a single function to compute recently played that takes a lambda to filter out the number of days since it's been played & call that twice with the two versions for < 1 week & > 1 week? It seems like that would avoid all the extra functions & Pair logic that was added here.

For example, something like:

```kotlin
  private fun computeRecentStoriesList(
    topicProgressList: List<TopicProgress>,
    completionTimeFilter: (Long) -> Boolean
  ): Pair<List<PromotedStory>, Long> {
    var numberOfDaysPassed = 0L
    val recentlyPlayedPromotedStoryList = mutableListOf<PromotedStory>()
    val sortedTopicProgressList =
      topicProgressList.sortedByDescending { it.lastPlayedTimestamp }.filter(completionTimeFilter)
...
```

And calls might look like:

```kotlin
val olderStories = computeRecentStoriesList(topicProgressList) { it > ONE_WEEK_IN_DAYS }
```

Reply:
Done.